### ISSUE TITLE
PS4 Games

### ISSUE THREAD

**Issue by AndaleObe** [paulov-t/FrostbiteModdingTool#77]
Hello,
I've been looking for a tool with which to edit Frostbite games for PS4 for a long time. I would like to extract textures from .toc .cas from NHL 22 PS4. is this possible with this tool? I would be happy about any reply.
Thank you

**Comment by AndaleObe** [comment]
I would have given you the files and everything you need--Diese Nachricht wurde von meinem Android Mobiltelefon mit GMX Mail gesendet.Am 03.10.23, 14:20 schrieb Paulov ***@***.***>:

   
  Is it possible? Yes. Will I be able to do it? No. I have always wanted NHL on PC but EA have never done it. PS4/PS5 files will require me to have the files to create a profile to do the work. Which I don't.

**Comment by AndaleObe** [comment]
can we exchange ideas where? discord or somewhere else?

**Comment by AndaleObe** [comment]
https://drive.google.com/file/d/1vS_L3WVozWXtPGI9GRcD6duPdL8whElM/view

This is a Patch from NHL 22 PS4. I have my Game dumped but is 40gb big.

**Comment by andresedu1996** [comment]
It is more feasible to do it for XBOX as it shares a store with Windows and you can release this app as UWP, however, I'm not sure if policies allow modding software on XBOX or its store. \'_'/

**Comment by AndaleObe** [comment]
It would be awesome if you could export texture and import it again.

**Comment by AndaleObe** [comment]
paulov, you wrote that ea can't get nhl out on the pc. That's exactly the reason why I started modding NHL for the PS4. I hate it when there are jerseys in the game that aren't right. especially in the European leagues. That's why I jailbroken my PS4 to mod NHL. I only use my NHL games. Up to NHL 21 it was still possible to mod, but from NHL 22 onwards with Frostbite it was no longer possible. But since the files like .toc .sb .cas look almost the same as on the PC, I thought to myself, since you are Frostbite experts, to extract textures, design them with Gimp Photoshop etc and put them back into the file. I would just need something to view these files and import - export. I then have to generate the files again as fpkg anyway in order to install them on the PS4. just the background why I'm looking for something like that.

**Comment by AndaleObe** [comment]
hey,
The two files are there where I sent the link, right?
or do you mean the two from the whole game?

**Comment by AndaleObe** [comment]
I uploaded to mediafire nhl 23 ps4. Everything is included except for the .cas files superbundles, of which I have two in the folder. I can't manage more with the gigabytes. Can I send you the link privately?

**Comment by AndaleObe** [comment]
https://1fichier.com/?hqcx8cqlp1keh0vhqkty&af=2899507

**Comment by AndaleObe** [comment]
This is the link to the complete PS4 NHL 23. If you have downloaded it, let me know briefly and I will upload a tool to my mediafire so that you can unpack the pkg, ok?

**Comment by AndaleObe** [comment]
does the link work?

**Comment by AndaleObe** [comment]
[NHL 22 PS4.txt](https://github.com/paulov-t/FrostbiteModdingTool/files/12816188/NHL.22.PS4.txt)

**Comment by AndaleObe** [comment]
Here is the Tools https://www.mediafire.com/folder/ev6nx4k1jf4pc/Neuer+Ordner+(2)

**Comment by AndaleObe** [comment]
Everything works?

**Comment by AndaleObe** [comment]
do you still need the link for the pkgeditor?

**Comment by AndaleObe** [comment]
see you already have some codes for NHL PS4 😃 you are a legend

**Comment by AndaleObe** [comment]
Have you been able to see textures in the editor?

**Comment by AndaleObe** [comment]
with the jerseys too? could it be that it is the .gnf format?

**Comment by AndaleObe** [comment]
PS4 used the .GNF format from NHL 16 to NHL 21. Is it possible that it will also be used in NHL 22?

**Comment by AndaleObe** [comment]
you could try exporting the file and opening it in noesis https://richwhitehouse.com/index.php?content=inc_projects.php. the tool supports the .gnf format

**Comment by AndaleObe** [comment]
OK, this is a .bin file. I don't know anything about .bin files 😒

**Comment by AndaleObe** [comment]
Are everything in FIFA also .bin files for PC?

**Comment by AndaleObe** [comment]
Then the hope of finally modding NHL 23 is over 😒

**Comment by AndaleObe** [comment]
I don't know who to ask. Maybe in the forum on discord for Skyrim PS4 or Battlefront PS4. I ask. Can I use and send your file if someone needs it?

**Comment by AndaleObe** [comment]
https://www.psxhax.com/threads/frost4-ps4-frostbite-engine-toolkit-gnmf-ba2-tools-by-socknastre.7285/

are you starting something?

**Comment by AndaleObe** [comment]
I managed it :)

**Comment by AndaleObe** [comment]
![grafik](https://github.com/paulov-t/FrostbiteModdingTool/assets/53269821/bc3f423b-a8a7-450e-8b66-9d11275d8af4)

**Comment by AndaleObe** [comment]
https://www.mediafire.com/folder/4dkfvsi0a234q/Neuer+Ordner

Open Rawtex.exe click on ps4 swizzle and select autodetected size 2048 and drag and drop the bin file. and now a dds file is converted

**Comment by AndaleObe** [comment]
The only question is, how do we get it back into the file?

**Comment by AndaleObe** [comment]
hey paulov-t,

did you download the files? then can I close the link again?

**Comment by AndaleObe** [comment]
Don't stress, when it's done, it's done. I've been looking for 2 years for someone to do this

**Comment by AndaleObe** [comment]
But just one question, do you think there is an import of the textures? and have you seen any legacy files like DB data like FIFA?

**Comment by AndaleObe** [comment]
For me personally I would only need the .db and the associated xml or .db language. everything else is not that interesting. and the textures. meshes files etc would be a treat but not necessary for me.

**Comment by AndaleObe** [comment]
I even believe that the Eboot.bin file represents the exe. I have this in hexedit. Maybe you could use these for sdk. just a thought 😊

**Comment by AndaleObe** [comment]
It is impertinent to ask you if you could send me jersey, pants and socks from Team Klagenfurter AC, EC Graz 99ers and Olimpija Ljubljana in advance. then I can edit it with gimp 😊

**Comment by AndaleObe** [comment]
yes, of course, I'll wait 😊 is it possible to import back into the file?

**Comment by AndaleObe** [comment]
I can't test until the afternoon 😃 Am I correct that I can export as a .bin file? so I have to convert a .dds with rawtex right?

**Comment by AndaleObe** [comment]
yes, of course that fits. If you have managed to import the .bin file back, the question is how do I get the .dds back into a .bin without a header. Do you have an idea in advance?

**Comment by AndaleObe** [comment]
hey,
how do I open nhl23? I click on the profile nhl23ps4 but I need an .exe

**Comment by AndaleObe** [comment]
OK, I did it. I created a folder with NHL23PS4, I inserted the data folder there and from the eboot.bin I renamed NHL23PS4.exe. everything worked

**Comment by AndaleObe** [comment]
If you could manage it, I could import it back and save the .cas or wherever the texture is. Then a dream would come true so that I could enjoy and watch it in the game. I was able to import the .db back without an error message, but saving it isn't possible yet

**Comment by AndaleObe** [comment]
hey paul,
If I can help in any way, please let me know

**Comment by AndaleObe** [comment]
Oh dear, unfortunately I'm too stupid for that. Unfortunately I know nothing about that.

**Comment by AndaleObe** [comment]
but if I understand correctly, FIFA 24 Switch has the same problems when exporting as NHL 24 PS4. you could install this rawtex tool, right?

**Comment by AndaleObe** [comment]
I converted the .bin texture to .dds with the rawtex tool. then I looked at the dds file in hexedit and found that the bin and dds are completely the same except that the dds has the header with dds and the bin doesn't. maybe you can add a code that adds the dds header?

**Comment by AndaleObe** [comment]
hey paul,
you did it 😃
Does that mean you can export the texture and import it back?
If you could save the whole thing now, I could try to generate the files for a pkg and see if everything works on the PS4

**Comment by AndaleObe** [comment]
@paulov-t 
Hey,
Are you perhaps releasing the tool over the weekend? 😃

**Comment by AndaleObe** [comment]
@paulov-t 
Unfortunately does not work. can not be opened. I created another NHL23PS4 folder. There I have the data folder and the eboot.bin which I renamed NHL24.exe. When I search for exe and select it, I get a very large error message in the tool

**Comment by AndaleObe** [comment]
![Unbenannt](https://github.com/paulov-t/FrostbiteModdingTool/assets/53269821/ee1c83ca-8249-496c-b9fe-7affd349ce7f)

**Comment by AndaleObe** [comment]
![Unbenannt](https://github.com/paulov-t/FrostbiteModdingTool/assets/53269821/a8a58d08-b1e4-474d-b9eb-f708995474f3)

**Comment by AndaleObe** [comment]
@paulov-t 
hey paul,
Do you also do something that can then be saved 😃? then I can try to gamble 😃❤️

**Comment by AndaleObe** [comment]
understood. but I can only export textures, nothing more, right?

**Comment by AndaleObe** [comment]
I exported the jerseys the first time and designed them with sponsors. Now I wanted to test what these look like in the game itself, but importing back and saving isn't possible yet, is it?

**Comment by AndaleObe** [comment]
@paulov-t 
I believe that the header for the textures is .gnf. This has been the case with all NHL PS4 parts so far. Can you manage to add a .gnf header for export and import? I have a .gnf to .dds and .dds to .gnf converter. If you still manage to save the .cas or .toc, I can try to see if it works on the PS4

**Comment by AndaleObe** [comment]
This is a jersey from NHL 21 PS4.

**Comment by AndaleObe** [comment]
[jersey_rbk_victoria_royals_three_cm.zip](https://github.com/paulov-t/FrostbiteModdingTool/files/13078768/jersey_rbk_victoria_royals_three_cm.zip)

**Comment by AndaleObe** [comment]
@paulov-t 
hey paul,
Legacy files are no longer displayed on NHL 23 PS4. I have the current FMT that you uploaded today

**Comment by AndaleObe** [comment]
@paulov-t 
May I ask if you have made any progress with import back texture and compiling?

**Comment by AndaleObe** [comment]
> Hi sorry no. i've been ill for the past week and just too tired to think about difficult things right now. i only released a quick something for fc24 yesterday to keep people off my back. now im just chilling :) maybe next week ill start?

I was sick too, but I'm getting better again. I have modified so many jerseys and it would be a dream to be able to see them on the PS4. Don't stress yourself out, health comes first. Once you've done it, import texture and save the bundles. then I'll press your donate button a little harder, believe me 😊

**Comment by AndaleObe** [comment]
@paulov-t 
I have the same problem as in discussion #66. but for me the sdk is not deleted or by antivirus. is normally present. The legacy editor still works with version FMT-24.0.8685.36478, but no longer works afterwards

**Comment by AndaleObe** [comment]
@paulov-t 
hey,
Are you starting to import the textures and save the files this week? 😊

**Comment by AndaleObe** [comment]
hey,
First of all, thank you for looking into this. I really appreciate that kind of thing. So for me, just importing back the Texture Jersey, Pants and Socks would be enough. And there only the _color file. No _coeff or anything else. In the legacy overview there is the file .db and .db xml. I could export them with your first version. that would be enough. Everything else like sound etc. is not important. If you can do it, it would be great and an extra treat 😁.

**Comment by AndaleObe** [comment]
I can do the testing. Unless you have a PS4 that you can jailbreak?

**Comment by AndaleObe** [comment]
@paulov-t 
Hey,
I see you've already accomplished something in yesterday's codes 😃

**Comment by AndaleObe** [comment]
Do you think I can test it out over the weekend? 😃

**Comment by AndaleObe** [comment]
> I havent finished the texture import yet, so no sorry 💀

Ok 😔

**Comment by AndaleObe** [comment]
hey,
With the new version of FMT the textures are no longer displayed. Have you changed anything?

**Comment by AndaleObe** [comment]
@paulov-t 
you managed to import the texture 😃

**Comment by AndaleObe** [comment]
hey,
saving doesn't work yet, does it? I can press game compile at the top right, then the bar appears and it says busy, but then nothing happens or the bar stays there forever.

**Comment by AndaleObe** [comment]
@paulov-t 
Maybe there would be a script where the files are simply extracted. quick bms would be nice, you could then simply do a reimport, right?

**Comment by AndaleObe** [comment]
Hey paul,
please don't give up on the project nhl 23 ps4. You already have the texture reimport. it would just save the .toc .sb .cas or whatever is needed.

**Comment by AndaleObe** [comment]
@paulov-t 
Now I was happy too early with the new FMT version. It's just an EAFC24 SDK update 😜

**Comment by AndaleObe** [comment]
@paulov-t 
please, please add the compiler for nhl 23.
so that the imported textures are also saved. 🙏🙏🙏

**Comment by AndaleObe** [comment]
![Unbenannt](https://github.com/paulov-t/FrostbiteModdingTool/assets/53269821/fc744df2-570c-4b88-b88d-555a3e52cdfb)

**Comment by AndaleObe** [comment]
I then tried to copy the profile from FC24 to the NHL 23 PS4. everything except "Name": "NHL23PS4",
  "DisplayName": "NHL23 [PS4]",
  "DataVersion": 20221011,
  "CacheName": "nhl23ps4", 

then I lanched and it even copies everything into a ModData folder. Unfortunately, only at the very end does the error message value was either to large or to small for an unsigned byte appear.

**Comment by AndaleObe** [comment]
hey paul,
is the code for compiling in NHL22PS4AssetCompiler.cs so that the changed .cas are saved?

**Comment by AndaleObe** [comment]
@paulov-t 
Hey,
please please can you compile for nhl 23 ps4? I've now modded all the leagues with jerseys, pants and socks and I'm done. I would love to see this in the game. please as a Christmas present 🙏🙏🙏🙏🙏

**Comment by AndaleObe** [comment]
@paulov-t 
I found something with which you can open the layout.toc and read out some data. maybe this will help you a little. There is also an InitFSCli.exe but I don't know what you can do with it. 
![Unbenannt](https://github.com/paulov-t/FrostbiteModdingTool/assets/53269821/4c6e6c76-f8b0-47b9-90c3-8ecc4542711b)

**Comment by AndaleObe** [comment]
[Frost4.Toolset.v1.0.0.0-beta.2.zip](https://github.com/paulov-t/FrostbiteModdingTool/files/13694868/Frost4.Toolset.v1.0.0.0-beta.2.zip)

**Comment by AndaleObe** [comment]
[Frost4-master.zip](https://github.com/paulov-t/FrostbiteModdingTool/files/13694879/Frost4-master.zip)

**Comment by AndaleObe** [comment]
i have add a cover for NHL23PS4 :)
![Unbenannt](https://github.com/paulov-t/FrostbiteModdingTool/assets/53269821/07cd3d83-900a-4b84-b76d-eaad581ebbac)

**Comment by AndaleObe** [comment]
@paulov-t 
Attached is the folder with the images if you would like to upload it
[images.zip](https://github.com/paulov-t/FrostbiteModdingTool/files/13751027/images.zip)

**Comment by AndaleObe** [comment]
hey @paulov-t 

Maybe you'll get back to work when you have time? 😊

**Comment by AndaleObe** [comment]
> > hey @paulov-t
> > Maybe you'll get back to work when you have time? 😊
> 
> I hope I can find some more time now I am finished with my other project 👍 . I did some signficant work to support compiling back to files a lot better than before. So this will be much better for console versions. Now gotta fix the texture import 👍

Glad you're back again 💪🏻

If you need anything, let me know 👍🏻

**Comment by AndaleObe** [comment]
Maybe you can also display the legacy files. So you had to use an old version where they were still displayed but the texture import didn't work. So one version for legacy files and another for texture. That was the case with NHL 23 😊
